### ISSUE TITLE
Compactor cleanup can use >10GB of Go heap

### ISSUE THREAD

**Issue by bboreham** [cortexproject/cortex#3794]
**Describe the bug**
Compactor 1.7.0-rc.0 OOM-crashlooping with a 10Gi limit.

I grabbed a heap profile a couple of seconds before it died:
[compactor-heap.pprof.gz](https://github.com/cortexproject/cortex/files/5932317/compactor-heap.pprof.gz)

This shows there is 4.38Gi in use on this stack:
```
Flat | Flat% | Sum% | Cum | Cum% | Name | Inlined?
4.38GB | 99.77% | 99.77% | 4.38GB | 99.80% | github.com/minio/minio-go/v7.Client.putObjectMultipartStreamNoLength |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/thanos-io/thanos/pkg/tracing.DoWithSpan |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/thanos-io/thanos/pkg/objstore/s3.(*Bucket).Upload |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/thanos-io/thanos/pkg/objstore.TracingBucket.Upload.func1 |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/thanos-io/thanos/pkg/objstore.TracingBucket.Upload |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/thanos-io/thanos/pkg/objstore.(*metricBucket).Upload |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/minio/minio-go/v7.Client.putObjectCommon |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/minio/minio-go/v7.Client.PutObject |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/cortexproject/cortex/pkg/util/concurrency.ForEachUser.func1 |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/cortexproject/cortex/pkg/storage/tsdb/bucketindex.MigrateBlockDeletionMarksToGlobalLocation |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/cortexproject/cortex/pkg/storage/tsdb/bucketindex.(*globalMarkersBucket).Upload |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/cortexproject/cortex/pkg/storage/bucket.(*UserBucketClient).Upload |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/cortexproject/cortex/pkg/compactor.(*BlocksCleaner).cleanUsers.func1 |  
0 | 0.00% | 99.77% | 4.38GB | 99.80% | github.com/cortexproject/cortex/pkg/compactor.(*BlocksCleaner).cleanUser |  
```

Source view:
```
github.com/minio/minio-go/v7.Client.putObjectMultipartStreamNoLength
/__w/cortex/cortex/vendor/github.com/minio/minio-go/v7/api-put-object.go

  Total:      4.38GB     4.38GB (flat, cum) 99.80%
    279            .          .            
    280            .          .           	// Initialize parts uploaded map. 
    281            .          .           	partsInfo := make(map[int]ObjectPart) 
    282            .          .            
    283            .          .           	// Create a buffer. 
    284       4.38GB     4.38GB           	buf := make([]byte, partSize) 
```

and by objects:
```
    284            7          7           	buf := make([]byte, partSize) 
```

so we have 7 blocks allocated totalling 4.38GB.

`partSize` is computed by some logic I didn't understand; I don't see why it should be in the hundreds of megabytes.

I can see from higher up the stack that `cleanUser` is running many users in parallel; reducing `-compactor.cleanup-concurrency` from the default 20 to 2 stopped the crash.

**Comment by bboreham** [comment]
Now I understand a bit more what this code is doing, my suspicion is that `putObjectMultipartStreamNoLength()` was allocating a 128MB buffer to upload a deletion marker of approximately 112 bytes.

Which I think was fixed by @pracucci in #3807.

**Comment by pracucci** [comment]
Yes, should be fixed by #3807. It's something we discovered just recently because this issue doesn't affect the GCS client (we use at Grafana Labs) so we learned it from community reports.

_Let me know if #3807 was enough or the issue is still there._
